mlos_bench/mlos_bench/storage/sql/experiment_data.py

```diff
@@ -52,56 +52,19 @@ def __init__(self, *,
-                        self._schema.objectives.c.optimization_target.asc(),
-                    )
+        with self._engine.connect() as conn:
+            objectives_db_data = conn.execute(
+                self._schema.objectives.select().where(
+                    self._schema.objectives.c.exp_id == self._experiment_id,
+                ).order_by(
+                    self._schema.objectives.c.weight.desc(),
+                    self._schema.objectives.c.optimization_target.asc(),
                 )
-                objectives = {
-                    objective.optimization_target: objective.optimization_direction
-                    for objective in objectives_db_data.fetchall()
-                }
-        # Backwards compatibility: try and obtain the objectives from the TrialData and merge them in.
-        # NOTE: The original format of storing opt_target/opt_direction in the Trial
-        # metadata did not support multi-objectives.
-        # Nor does it make it easy to detect when a config change caused a switch in
-        # opt_direction for a given opt_target between run.py executions of an
-        # Experiment.
-        # For now, we simply issue a warning about potentially inconsistent data.
-        for trial in self.trials.values():
-            trial_objs_df = trial.metadata_df[
-                trial.metadata_df["parameter"].isin(("opt_target", "opt_direction"))
-            ][["parameter", "value"]]
-            try:
-                opt_targets = trial_objs_df[trial_objs_df["parameter"] == "opt_target"]
-                assert len(opt_targets) == 1, \
-                    "Should only be a single opt_target in the metadata params."
-                opt_target = opt_targets["value"].iloc[0]
-            except KeyError:
-                continue
-            try:
-                opt_directions = trial_objs_df[trial_objs_df["parameter"] == "opt_direction"]
-                assert len(opt_directions) <= 1, \
-                    "Should only be a single opt_direction in the metadata params."
-                opt_direction = opt_directions["value"].iloc[0]
-            except (KeyError, IndexError):
-                opt_direction = None
-            if opt_target not in objectives:
-                objectives[opt_target] = opt_direction
-            elif opt_direction != objectives[opt_target]:
-                _LOG.warning("Experiment %s has multiple trial optimization directions for optimization_target %s=%s",
-                             self, opt_target, objectives[opt_target])
-        for opt_tgt, opt_dir in objectives.items():
-            assert opt_dir in {None, "min", "max"}, f"Unexpected opt_dir {opt_dir} for opt_tgt {opt_tgt}."
-        return objectives
+            )
+            return {
+                objective.optimization_target: objective.optimization_direction
+                for objective in objectives_db_data.fetchall()
+            }
```

Comment:
This was tricky to write when first implemented.  Should make sure we have some good test coverage for it (may already be there, but worth double checking).

Reply:
I would not rely on trials' metadata to determine the experiment's objective. To me, the Experiment's objectives are defined when we create the experiment; if later we merge in some other trials into it, the experiment's objectives stay the same, even if some trials had been run with different objectives or no objectives at all. The info on individual trial's objectives we can get from the metadata, and I will soon implement an accessor `TrialData.objectives` to facilitate that (issue #729)

Reply:
I agree on the first part.  Using metadata was a means to backport functionality since early on we didn't have a separate `objectives` table that could be used to pull the data from the `Experiment` directly.

But I don't think we should expose the objectives for TrialData itself - its less relevant there versus `Experiment` for exactly the merging operation you point out.